services/report/__init__.py

```diff
@@ -1173,6 +1191,74 @@ def save_report(self, commit: Commit, report: Report, report_code=None):
         )
         return {"url": url}
 
+    def _determine_cffs_and_depths_in_db(
+        self, db_session, commit: Commit
+    ) -> Optional[Dict[int, int]]:
+        depths_with_cff_upload_ids = defaultdict(list)
+
+        existing_cff_uploads: Optional[List[Upload]] = (
+            db_session.query(Upload)
+            .filter_by(
+                report_id=commit.report.id_,
+                upload_type=SessionType.carriedforward.value,
+            )
+            .all()
+        )
+
+        if existing_cff_uploads:
+            for upload in existing_cff_uploads:
+                parent_depth = upload.upload_extras.get("parent_depth")
+                if parent_depth:
+                    depths_with_cff_upload_ids[parent_depth].append(upload.id)
+
+            log.info(
+                "Existing cffs and their respective depths",
+                extra=dict(
+                    commit=commit.commitid,
+                    repo=commit.repoid,
+                    existing_uploads=existing_cff_uploads,
+                ),
+            )
+
+        return depths_with_cff_upload_ids
+
+    def _possibly_delete_existing_cffs(
+        self,
+        db_session,
+        commit: Commit,
+        carryforward_report: Report,
+    ):
+        cffs_and_depths_in_db = self._determine_cffs_and_depths_in_db(
+            db_session=db_session, commit=commit
+        )
+
+        if not cffs_and_depths_in_db:
+            return
+
+        # Gets first 'parent_depth' it finds, as all cffs should be carried forward
+        # by the same parent within the same report
+        carryforward_report_depth = MIN_PARENT_COMMIT_DEPTH
+        for upload in carryforward_report.sessions.values():
+            if upload.session_extras.get("parent_depth"):
+                carryforward_report_depth = upload.session_extras.get("parent_depth")
+                break
+
+        # Delete all cff uploads that have a higher depth than the one in
+        # the latest report
+        for parent_depth, upload_ids in cffs_and_depths_in_db.items():
+            if parent_depth > carryforward_report_depth:
+                log.info(
+                    "Deleting upload from DB",
+                    extra=dict(
+                        repoid=commit.repoid,
+                        commit=commit.commitid,
+                    ),
+                )
+                # Potentially make task here - this would make the preprocess task likely timeout
```

Comment:
The upload task would be likely to timeout here too? Because that's a way bigger problem

Reply:
It could depending on the amount of things to delete, hence the ask here to see if it would make sense to create a separate deletion task

Reply:
I think that would be a good idea, yes.
Especially if it only changes the database and not the `Report` object it would make a lot of sense to move this to a different task